In [ ]:
# Install MySQL Server, ipython-sql, and pymysql
!apt-get update

Hit:1 http://archive.ubuntu.com/ubuntu jammy InRelease
Get:2 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [128 kB]
Get:3 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,632 B]
Get:4 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease [1,581 B]
Get:5 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]
Get:6 https://r2u.stat.illinois.edu/ubuntu jammy InRelease [6,555 B]
Get:7 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [127 kB]
Hit:8 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease
Hit:9 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease
Get:10 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease [24.6 kB]
Get:11 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  Packages [1,683 kB]
Get:12 http://archive.ubuntu.com/ubuntu jammy-updates/restricted amd64 Packages [4,517 kB]
Get:13 https://r2u.stat.illi

In [ ]:
!apt-get install -y mysql-server

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
  libcgi-fast-perl libcgi-pm-perl libclone-perl libencode-locale-perl
  libfcgi-bin libfcgi-perl libfcgi0ldbl libhtml-parser-perl
  libhtml-tagset-perl libhtml-template-perl libhttp-date-perl
  libhttp-message-perl libio-html-perl liblwp-mediatypes-perl libmecab2
  libprotobuf-lite23 liburi-perl mecab-ipadic mecab-ipadic-utf8 mecab-utils
  mysql-client-8.0 mysql-client-core-8.0 mysql-server-8.0
  mysql-server-core-8.0
Suggested packages:
  libdata-dump-perl libipc-sharedcache-perl libbusiness-isbn-perl libwww-perl
  mailx tinyca
The following NEW packages will be installed:
  libcgi-fast-perl libcgi-pm-perl libclone-perl libencode-locale-perl
  libfcgi-bin libfcgi-perl libfcgi0ldbl libhtml-parser-perl
  libhtml-tagset-perl libhtml-template-perl libhttp-date-perl
  libhttp-message-perl libio-html-perl liblwp-mediatypes-perl libmecab2
  lib

In [ ]:
!pip install ipython-sql pymysql

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.0/45.0 kB 3.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 29.5 MB/s eta 0:00:00


In [ ]:
# Start MySQL server
!service mysql start




 * Starting MySQL database server mysqld
su: warning: cannot change directory to /nonexistent: No such file or directory
   ...done.


In [ ]:
# Set MySQL root password and create the database
!mysql -e "ALTER USER 'root'@'localhost' IDENTIFIED WITH mysql_native_password BY 'root'; FLUSH PRIVILEGES;"

In [ ]:
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


In [ ]:
!pip install ipython-sql PyMySQL SQLAlchemy==1.4.49 prettytable==3.6.0 --quiet


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 33.4 MB/s eta 0:00:00


In [ ]:
# Create the database
!mysql -u root -proot -e "CREATE DATABASE IF NOT EXISTS adashi_staging;"



mysql: [Warning] Using a password on the command line interface can be insecure.


In [ ]:
# Import the SQL file into the new database
!mysql -u root -proot adashi_staging < /content/drive/MyDrive/adashi_assessment.sql


mysql: [Warning] Using a password on the command line interface can be insecure.


In [ ]:
# Load the ipython-sql extension
%load_ext sql




The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [ ]:
# Connect to the adashi_staging database
%sql mysql+pymysql://root:root@localhost/adashi_staging

In [ ]:
import prettytable
from prettytable import TableStyle

# Set default table style using TableStyle enum
if not hasattr(prettytable, 'DEFAULT'):
    prettytable.DEFAULT = TableStyle.MSWORD_FRIENDLY


In [ ]:
# Show all tables
%sql SHOW TABLES;

 * mysql+pymysql://root:***@localhost/adashi_staging
4 rows affected.


[('plans_plan',),
 ('savings_savingsaccount',),
 ('users_customuser',),
 ('withdrawals_withdrawal',)]

In [ ]:
%%sql
SELECT 'plans_plan' AS table_name, COUNT(*) AS row_count FROM plans_plan
UNION ALL
SELECT 'savings_savingsaccount', COUNT(*) FROM savings_savingsaccount
UNION ALL
SELECT 'users_customuser', COUNT(*) FROM users_customuser
UNION ALL
SELECT 'withdrawals_withdrawal', COUNT(*) FROM withdrawals_withdrawal;


 * mysql+pymysql://root:***@localhost/adashi_staging
4 rows affected.


[('plans_plan', 9641),
 ('savings_savingsaccount', 163736),
 ('users_customuser', 1867),
 ('withdrawals_withdrawal', 1308)]

In [ ]:
%%sql
SELECT
    t.table_name,
    IFNULL(r.row_count, 0) AS row_count,
    COUNT(c.column_name) AS column_count
FROM
    information_schema.tables t
LEFT JOIN (
    SELECT
        'plans_plan' AS table_name, COUNT(*) AS row_count FROM plans_plan
    UNION ALL
    SELECT 'savings_savingsaccount', COUNT(*) FROM savings_savingsaccount
    UNION ALL
    SELECT 'users_customuser', COUNT(*) FROM users_customuser
    UNION ALL
    SELECT 'withdrawals_withdrawal', COUNT(*) FROM withdrawals_withdrawal
) r ON t.table_name = r.table_name
JOIN information_schema.columns c
    ON t.table_name = c.table_name
WHERE
    t.table_schema = 'adashi_staging'
    AND t.table_name IN ('plans_plan', 'savings_savingsaccount', 'users_customuser', 'withdrawals_withdrawal')
GROUP BY
    t.table_name, r.row_count;


 * mysql+pymysql://root:***@localhost/adashi_staging
4 rows affected.


[('plans_plan', 9641, 56),
 ('savings_savingsaccount', 163736, 37),
 ('users_customuser', 1867, 57),
 ('withdrawals_withdrawal', 1308, 23)]

In [89]:
#Users with no savings activity
%%sql
SELECT u.id, u.email
FROM users_customuser u
LEFT JOIN savings_savingsaccount s ON u.id = s.owner_id
WHERE s.id IS NULL;


 * mysql+pymysql://root:***@localhost/adashi_staging
994 rows affected.


[('aae6bc85b49d4e4aa6d625738e7bd989', 'a@admin.com'),
 ('8ff0c8a0d3ab4c3a9ee7fb58447591ea', 'aclpool+james@gmail.com'),
 ('f35a6dec41eb43aab5bd924ff48702b6', 'aclpool+koko11@cowrywise.com'),
 ('d920e45a23c345e9bb92c17aba85b42d', 'aclpool+koko12@cowrywise.com'),
 ('aea9a6163b2042baaacb78070bcf8cf9', 'aclpool+koko13@cowrywise.com'),
 ('ddd26ea905d94b819bb7325d8d183f2b', 'aclpool+koko14@cowrywise.com'),
 ('6363c58649b04224ba4eae4628b4a35b', 'aclpool+koko15@cowrywise.com'),
 ('55d188417f424459ba3222bbb90882ba', 'aclpool+master@gmail.com'),
 ('a5bf3dbcf8674fe994e42d8ab9b81e8f', 'aclpool+segun@gmail.com'),
 ('ae53a225ca9a4dbbb7834ac6cedec154', 'aclpool+testnum@cowrywise.com'),
 ('6504479ec51641b9bfe608ed7c3adad3', 'aclpool+threes@gmail.comxxxxxxxxx'),
 ('2a8524fa8a7f4a46b296b5eb3dc5b1d8', 'adimchi+20@cowrywise.com'),
 ('0c8bc6dcf1d94f93975a399c1784ad24', 'adimchi+2@cowrywise.com'),
 ('d0060a3ae5c441559b84da37dc8c3fbb', 'adimchi+3@cowrywise.com'),
 ('576530ee0ebf4585a506e8fcc89ecd25', 'adimchi+6@cowrywise.com'),
 ('bfadd92623ad44e98eb00978f03c4205', 'afoo+test10@cowrywise.com'),
 ('3f9c10f4f34e4c12b79e131c8f8219fe', 'afoo+test11@cowrywise.com'),
 ('75061e844ecd40f5a8c1096c2df8edf0', 'afoo+test12@cowrywise.com'),
 ('1d1e1dc8685945e2b2dddc49803e2dcb', 'afoo+test13@cowrywise.com'),
 ('a94ce5891733488a9428c9a1d1254172', 'afoo+test14@cowrywise.com'),
 ('acd06f68a88947c88786a8d52e08202e', 'afoo+test6@cowrywise.com'),
 ('655cf5b1155b46aba2e0a1c3299b01c8', 'afoo+test7@cowrywise.com'),
 ('806fba4574484bc9b5d488aa48f4320d', 'afoo+test9@cowrywise.com'),
 ('11af8c455bf747a49e7115425f84c5ee', 'afoo@cowrywise.com'),
 ('57bb571ce4ff441ab251f3b14ad37c57', 'ahmed+1000+daniella@cowrywise.com'),
 ('f25780b3cfa34dae9252083db744940a', 'ahmed+1000+david@cowrywise.com'),
 ('96c55ec2cb7e406d9bed0f24a7e4bca8', 'ahmed+1000+eliana@cowrywise.com'),
 ('70147a2b69bc44ba936442951271caa0', 'aisha+11@cowrywise.com'),
 ('aba81eb463484cbfa8cf5d754406d47c', 'aisha+1@cowrywise.com'),
 ('95f4a650ebed4fd0a4d630eebf6e9f1e', 'aisha+2@cowrywise.com'),
 ('3b29d0ddf3b64a51b5688510eb7ca45f', 'aisha+3@cowrywise.com'),
 ('c03820b795bb4383a98fb298601e8d09', 'anagboguchinonso97@gmail.com'),
 ('86e5ff30c74b464bae482a1b7efdc043', 'anagboguchristopher@gmail.com'),
 ('569a2e67ed4c42aab26312a2422b846c', 'anakin@cowrywise.com'),
 ('156dcb9f229d46b3bec1802c14094b0b', 'aquila4christ@gmail.com'),
 ('a3c05cca7dda444ba3fb19072116620d', 'ayo16345@gmail.com'),
 ('819278f49937410e986577e2218486d2', 'b@admin.com'),
 ('591ef0168af044b6838e07a491b807bc', 'barakat@cowrywise.com'),
 ('d561bd5987ad4b3a9bd99829b73ea102', 'blessing+test@cowrywise.com'),
 ('46d74027d6a044909f900ddf623cf58d', 'blessing@cowrywise.com'),
 ('8265e84236f54847a2daccadbd8517e1', 'boluwatifesamuel48@gmail.com'),
 ('fba7f27812e4425698d06f08ce99b3d9', 'c@admin.com'),
 ('bb7c9d604fa64dcbb7070bb082c10956', 'chigozie+aaron@cowrywise.com'),
 ('4b671eb5677e4c808b8ed1f756e87532', 'chima+99@cowrywise.com'),
 ('d3edd528c8c540aeb7fa0ff5a7ed8d91', 'chima+ads@cowrywise.com'),
 ('57774e330e39468d9e8f1a6169ffbeb2', 'chima+super@cowrywise.com'),
 ('1c5dbc333ab3442e97daf332c20abc30', 'chinonso+1009@cowrywise.com'),
 ('dabd9b95106942d7bd9fe0a236eb1aa4', 'chinonso+100@cowrywise.com'),
 ('bbd13fb280a544eab39ec89559928afd', 'chinonso+101@cowrywise.com'),
 ('8699bb479c8641199957cf0ca2a5aae9', 'chinonso+102@cowrywise.com'),
 ('7bb2bb3761af4ef5b8d54ed0c21620d5', 'chinonso+103@cowrywise.com'),
 ('d88641bd4b274ed3ba11f330ef50758a', 'chinonso+104@cowrywise.com'),
 ('267a2e5b5a8b463ca7ff5ed55370262d', 'chinonso+105@cowrywise.com'),
 ('98c822c89d5e4bbc8f4d0cc5fc73679f', 'chinonso+108@cowrywise.com'),
 ('b09d5ed40c164a0ca4f7d6ae4c442d96', 'chinonso+110@cowrywise.com'),
 ('0df01f83e7bd4b09ac2cdba7869d93fc', 'chinonso+111@cowrywise.com'),
 ('196105f43a29424d924d73bb3d741eb2', 'chinonso+112@cowrywise.com'),
 ('335724a35af24535acd98913ca58d0f7', 'chinonso+113@cowrywise.com'),
 ('a66022753bb046a9b0a6b69266a59608', 'chinonso+114@cowrywise.com'),
 ('d1bd13ec7b9846e

In [ ]:
%%sql
SELECT * FROM plans_plan LIMIT 5;



 * mysql+pymysql://root:***@localhost/adashi_staging
5 rows affected.


[('000cbd97d4874dad86eef923db578da8', 'Stash', 'Stash', 0, datetime.date(2023, 10, 28), None, None, datetime.datetime(2023, 10, 28, 12, 30, 19), 1, '27cbc86d9dd0474c8b32634999aed816', 2, 0.0, None, 0, 1, 0.0, 1, datetime.date(2023, 10, 28), datetime.date(2023, 10, 28), 0, None, 0, 0, 0, 0, 0, None, 0, 0, None, 1, 0, 0, 0, 'COWRYWISE', 0, '', None, None, None, 0, 0, 0, 0, 1, 0, None, 0, 0, None, 0, 0, None, 0, 0, None),
 ('001357492be14d51bb15e4a713686d89', 'Stash', 'Stash', 0, datetime.date(2022, 3, 2), None, None, datetime.datetime(2022, 3, 2, 9, 38, 31), 1, '35851a03131d4b46837acc586021d538', 2, 0.0, None, 0, 1, 0.0, 1, datetime.date(2022, 3, 2), datetime.date(2022, 3, 2), 0, 'cac7bc258d17406587efcd10a0588381', 0, 0, 0, 0, 0, None, 0, 0, None, 1, 0, 0, 0, 'COWRYWISE', 0, '', None, None, None, 0, 0, 0, 0, 1, 0, None, 0, 0, None, 0, 0, None, 0, 0, None),
 ('002b48c9f6ec48fdb586bd019a85aa9a', 'Emergency', 'Regular Savings', 2045300, datetime.date(2022, 4, 14), datetime.date(2025, 4, 17), datetime.date(2025, 4, 24), datetime.datetime(2022, 4, 14, 13, 2, 48), 2, '5572810f38b543429ffb218ef15243fc', 1, 0.41, None, 0, 4, 409048500.0, 0, datetime.date(2025, 4, 20), datetime.date(2025, 4, 19), 0, 'faa7339082a7469aabffb16ceb0810a0', 0, 0, 0, 0, 0, None, 0, 0, None, 0, 0, 0, 0, 'COWRYWISE', 0, '', None, None, None, 1, 0, 1, 0, 1, 0, None, 0, 0, None, 0, 0, None, 0, 0, 1),
 ('002de81fe90549a9bd545c57b10a8007', 'ollp', 'Group', 333300, datetime.date(2023, 10, 20), datetime.date(2025, 4, 19), datetime.date(2025, 4, 20), datetime.datetime(2023, 10, 20, 12, 22), 6, '3aa79f2f1c0148cd964a6f91dd0fd72b', 1, 0.11, None, 0, 1, 0.0, 0, datetime.date(2025, 4, 20), datetime.date(2025, 4, 19), 0, '01f3e959a4264363b326625094227db7', 0, 0, 0, 0, 0, 'a8941b0353c742ea9856cb276276a7f3', 0, 0, None, 0, 0, 0, 0, 'COWRYWISE', 0, '', None, None, None, 1, 0, 0, 0, 1, 0, None, 0, 0, 'DTSTART:20231020T000000Z\nRRULE:FREQ=DAILY', 0, 0, None, 0, 0, None),
 ('00307f85a85c4457bc558db03e8e4778', 'Stash', 'Stash', 0, datetime.date(2024, 12, 17), None, None, datetime.datetime(2024, 12, 17, 9, 27, 11), 1, 'a058bcd7910d454395e62102e798f0f1', 2, 0.0, None, 0, 1, 0.0, 1, datetime.date(2024, 12, 17), datetime.date(2024, 12, 17), 0, None, 0, 0, 0, 0, 0, None, 0, 0, None, 1, 0, 0, 0, 'COWRYWISE', 0, '', None, None, None, 0, 0, 0, 0, 1, 0, None, 0, 0, None, 0, 0, None, 0, 0, None)]

In [ ]:
#Total number of plans and summary
%%sql
SELECT
    COUNT(*) AS total_plans,
    COUNT(DISTINCT owner_id) AS total_unique_users,
    AVG(amount) AS avg_plan_amount,
    SUM(amount) AS total_amount
FROM plans_plan;


 * mysql+pymysql://root:***@localhost/adashi_staging
1 rows affected.


[(9641, 1739, Decimal('8945667.6233'), Decimal('86245181556'))]

In [ ]:
#Count of plans by name (type)
%%sql

SELECT
    name,
    COUNT(*) AS count
FROM plans_plan
GROUP BY name
ORDER BY count DESC
LIMIT 10;


 * mysql+pymysql://root:***@localhost/adashi_staging
10 rows affected.


[('Stash', 1566),
 ('Dollar Stash', 1504),
 ('Emergency', 419),
 ('General', 289),
 ('First plan', 142),
 ('Own A Home', 86),
 ('A New Car', 72),
 ('Conservative Portfolio', 68),
 ('Conservative Portfolio 2', 67),
 ('Have a Baby', 60)]

In [ ]:
#Distribution of active vs deleted/archived plans
%%sql

SELECT
    is_archived,
    is_deleted,
    COUNT(*) AS count
FROM plans_plan
GROUP BY is_archived, is_deleted;


 * mysql+pymysql://root:***@localhost/adashi_staging
2 rows affected.


[(0, 0, 8980), (0, 1, 661)]

is_archived = 0 and is_deleted = 0 → 8980 plans
These are active plans. They are neither archived nor deleted, meaning they're likely in use or available to the user.

is_archived = 0 and is_deleted = 1 → 661 plans
These are soft-deleted plans. They're not archived, but are marked as deleted usually this means the record still exists in the database but is hidden from the user in the UI.

In [ ]:
%%sql
DESCRIBE users_customuser;


 * mysql+pymysql://root:***@localhost/adashi_staging
57 rows affected.


[('password', 'varchar(128)', 'NO', '', None, ''),
 ('last_login', 'datetime', 'YES', '', None, ''),
 ('is_superuser', 'tinyint(1)', 'NO', '', None, ''),
 ('id', 'char(32)', 'NO', 'PRI', None, ''),
 ('email', 'varchar(60)', 'NO', 'UNI', None, ''),
 ('name', 'varchar(100)', 'YES', '', None, ''),
 ('first_name', 'varchar(100)', 'YES', '', None, ''),
 ('last_name', 'varchar(100)', 'YES', '', None, ''),
 ('phone_number', 'varchar(128)', 'YES', '', None, ''),
 ('date_of_birth', 'date', 'YES', '', None, ''),
 ('is_staff', 'tinyint(1)', 'NO', '', None, ''),
 ('is_active', 'tinyint(1)', 'NO', '', None, ''),
 ('date_joined', 'datetime', 'NO', '', None, ''),
 ('is_admin', 'tinyint(1)', 'NO', '', None, ''),
 ('username', 'varchar(25)', 'YES', '', None, ''),
 ('created_on', 'datetime', 'NO', '', None, ''),
 ('gender_id', 'int', 'YES', 'MUL', None, ''),
 ('invite_code', 'varchar(100)', 'YES', '', None, ''),
 ('avatar_firebase_reference', 'varchar(255)', 'YES', '', None, ''),
 ('avatar_local_uri', 'varchar(255)', 'YES', '', None, ''),
 ('avatar_url', 'varchar(255)', 'YES', '', None, ''),
 ('risk_apetite', 'int', 'NO', '', None, ''),
 ('current_latitude', 'varchar(100)', 'YES', '', None, ''),
 ('current_longitude', 'varchar(100)', 'YES', '', None, ''),
 ('postal_address', 'varchar(255)', 'YES', '', None, ''),
 ('pin', 'varchar(128)', 'YES', '', None, ''),
 ('ambassador_profile_id', 'char(32)', 'YES', 'UNI', None, ''),
 ('is_ambassador', 'tinyint(1)', 'NO', '', None, ''),
 ('account_source', 'varchar(100)', 'NO', '', None, ''),
 ('is_vendor_account', 'tinyint(1)', 'NO', '', None, ''),
 ('is_business_account', 'tinyint(1)', 'NO', '', None, ''),
 ('is_account_deleted', 'tinyint(1)', 'NO', '', None, ''),
 ('vendor_code', 'varchar(100)', 'NO', '', None, ''),
 ('is_halal_account', 'tinyint(1)', 'NO', '', None, ''),
 ('address_city', 'varchar(100)', 'YES', '', None, ''),
 ('address_country', 'varchar(40)', 'YES', '', None, ''),
 ('address_state', 'varchar(40)', 'YES', '', None, ''),
 ('address_street', 'varchar(255)', 'YES', '', None, ''),
 ('monthly_expense', 'double', 'NO', '', None, ''),
 ('monthly_salary', 'double', 'NO', '', None, ''),
 ('is_account_disabled', 'tinyint(1)', 'NO', '', None, ''),
 ('authy_id', 'varchar(128)', 'YES', '', None, ''),
 ('fraud_score', 'int', 'NO', '', None, ''),
 ('account_campaign', 'varchar(150)', 'YES', '', None, ''),
 ('account_medium', 'varchar(100)', 'YES', '', None, ''),
 ('last_password_change', 'datetime(6)', 'YES', '', None, ''),
 ('last_pin_change', 'datetime(6)', 'YES', '', None, ''),
 ('is_private', 'tinyint(1)', 'NO', '', None, ''),
 ('disabled_at', 'datetime(6)', 'YES', '', None, ''),
 ('is_disabled_by_owner', 'tinyint(1)', 'NO', '', None, ''),
 ('is_account_deleted_by_owner', 'tinyint(1)', 'NO', '', None, ''),
 ('proposed_deletion_date', 'datetime(6)', 'YES', '', None, ''),
 ('reason_for_deletion', 'varchar(255)', 'YES', '', None, ''),
 ('enabled_at', 'datetime(6)', 'YES', '', None, ''),
 ('signup_device', 'varchar(100)', 'YES', '', None, ''),
 ('proposed_enablement_date', 'datetime(6)', 'YES', '', None, ''),
 ('tier_id', 'int', 'YES', 'MUL', None, '')]

In [ ]:
#Shows top 10 users with the most deleted plans.

%%sql
SELECT
    owner_id,
    COUNT(*) AS deleted_plans_count
FROM
    plans_plan
WHERE
    is_deleted = 1
GROUP BY
    owner_id
ORDER BY
    deleted_plans_count DESC
LIMIT 10;

 * mysql+pymysql://root:***@localhost/adashi_staging
10 rows affected.


[('0257625a02344b239b41e1cbe60ef080', 146),
 ('b70eddd82eae4684bddc252b0b05e895', 97),
 ('72141b6db0a94e9b8414ae0e783792b7', 65),
 ('279602fb41c240689c5edfec63c7c1a1', 57),
 ('a549807eb71046c48857f1a32277bd71', 51),
 ('6e52f5ecbf194634aedca49ba39a8a19', 33),
 ('5572810f38b543429ffb218ef15243fc', 19),
 ('26170a1ffd6f4b1f920da32dca5ae88e', 16),
 ('cccfe6a3d67b4a80b51e80be3e33eb5d', 14),
 ('b15810c73f1f4722b2ea72d9e329ea0c', 13)]

In [ ]:
#Group deleted plans by year, months of created_on
%%sql
SELECT
    DATE_FORMAT(created_on, '%Y-%m') AS created_month,
    COUNT(*) AS deleted_count
FROM
    plans_plan
WHERE
    is_deleted = 1
GROUP BY
    created_month
ORDER BY
    deleted_count DESC
LIMIT 10;

 * mysql+pymysql://root:***@localhost/adashi_staging
10 rows affected.


[('2021-05', 61),
 ('2021-11', 39),
 ('2018-06', 38),
 ('2019-10', 36),
 ('2022-05', 34),
 ('2021-07', 33),
 ('2020-01', 29),
 ('2019-09', 23),
 ('2022-03', 20),
 ('2018-02', 16)]

In [ ]:
%%sql
SELECT
    DATE_FORMAT(last_charge_date, '%Y-%m') AS charge_month,
    COUNT(*) AS deleted_count
FROM
    plans_plan
WHERE
    is_deleted = 1 AND last_charge_date IS NOT NULL
GROUP BY
    charge_month
ORDER BY
    deleted_count DESC
LIMIT 10;

 * mysql+pymysql://root:***@localhost/adashi_staging
10 rows affected.


[('2021-07', 14),
 ('2021-06', 9),
 ('2018-03', 7),
 ('2018-06', 6),
 ('2018-02', 6),
 ('2020-04', 6),
 ('2017-10', 5),
 ('2017-11', 4),
 ('2019-10', 4),
 ('2020-12', 4)]

In [ ]:
%%sql SELECT * FROM savings_savingsaccount LIMIT 5;

 * mysql+pymysql://root:***@localhost/adashi_staging
5 rows affected.


[(1, '1423f34658a34b05a34583daeb4a176b', datetime.datetime(2016, 8, 28, 21, 56, 43), datetime.datetime(2016, 9, 4, 21, 56, 43), 10000.0, 10000.0, 0.0, 10000.0, datetime.datetime(2016, 8, 28, 21, 56, 43), 'T40046006', 'success', '10000', 'Verification successful', '200', datetime.datetime(2016, 8, 28, 21, 56, 45), 4.38356164383562, 0.0, 0.0, datetime.date(2016, 9, 5), datetime.date(2016, 9, 6), datetime.datetime(2016, 8, 28, 21, 56, 43), None, 1, 1, '0257625a02344b239b41e1cbe60ef080', '671e6e5caf0f463b80123c839264de8c', 1, 1, '', 0.0, None, 0, '', None, None, None, 0.0),
 (4, '87a768a9a08c47d79c92bcd4e08a6c61', datetime.datetime(2016, 9, 1, 0, 37, 43), datetime.datetime(2016, 9, 1, 0, 37, 43), 10000.0, 10000.0, 0.0, 30004.3835616438, datetime.datetime(2016, 9, 1, 0, 37, 43), 'T03148197', 'success', '10000', 'Verification successful', '200', datetime.datetime(2016, 9, 1, 0, 38, 1), 8.21917808219178, 8.21917808219178, 0.0, datetime.date(2016, 9, 5), datetime.date(2016, 9, 6), datetime.datetime(2016, 9, 1, 0, 37, 43), None, 1, 1, '0257625a02344b239b41e1cbe60ef080', '671e6e5caf0f463b80123c839264de8c', 1, 1, '', 0.0, None, 0, '', None, None, None, 0.0),
 (13, '279a5bc39fde4faf943d1eadef9ded1e', datetime.datetime(2016, 9, 18, 19, 7, 14), datetime.datetime(2016, 9, 18, 19, 7, 14), 10000.0, 10000.0, 10000.0, 10000.0, datetime.datetime(2016, 9, 18, 19, 7, 14), 'T64001585', 'success', '10000', 'Verification successful', '200', datetime.datetime(2016, 9, 18, 19, 7, 32), 0.0, 0.0, 0.0, datetime.date(2016, 9, 18), datetime.date(2016, 9, 20), datetime.datetime(2016, 9, 18, 19, 7, 14), None, 1, 1, '0257625a02344b239b41e1cbe60ef080', 'ba6cda07fcd44c6a812fb0b7fee01b3c', 5, 1, '', 0.0, None, 0, '', None, None, None, 0.0),
 (14, '116437619f414c2990ac3c221d449346', datetime.datetime(2016, 9, 18, 19, 21, 49), datetime.datetime(2016, 9, 18, 19, 21, 49), 10000.0, 10000.0, 10000.0, 10000.0, datetime.datetime(2016, 9, 18, 19, 21, 49), 'T88173951', 'success', '10000', 'Verification successful', '200', datetime.datetime(2016, 9, 18, 19, 22, 18), 0.0, 0.0, 0.0, datetime.date(2016, 9, 18), datetime.date(2016, 9, 20), datetime.datetime(2016, 9, 18, 19, 21, 49), None, 1, 1, '0257625a02344b239b41e1cbe60ef080', 'dc58f3ec56e04193b818a473b3ae8ff4', 5, 1, '', 0.0, None, 0, '', None, None, None, 0.0),
 (18, '963c8ffaf86f4929b35600ecdf965887', datetime.datetime(2016, 9, 26, 0, 58, 42), datetime.datetime(2016, 9, 26, 0, 58, 42), 10000.0, 10000.0, 10000.0, 10000.0, datetime.datetime(2016, 9, 26, 0, 58, 42), 'T14225855', 'success', '10000', 'Verification successful', '200', datetime.datetime(2016, 9, 26, 0, 59, 31), 21.9178082191782, 21.9178082191782, 0.0, datetime.date(2016, 10, 5), datetime.date(2016, 10, 6), datetime.datetime(2016, 9, 26, 0, 58, 42), None, 1, 1, '0257625a02344b239b41e1cbe60ef080', '2a090fbd86984e6f9b4d03c0badcb866', 1, 1, '', 0.0, None, 0, '', None, None, None, 0.0)]

In [ ]:

#Total saved amount and average maturity period (days)
%%sql
SELECT
    COUNT(*) AS total_accounts,
    SUM(amount) AS total_amount_saved,
    AVG(DATEDIFF(maturity_end_date, maturity_start_date)) AS avg_maturity_days
FROM
    savings_savingsaccount
WHERE
    transaction_status = 'success';

 * mysql+pymysql://root:***@localhost/adashi_staging
1 rows affected.


[(106727, 470196161061.3264, Decimal('0.0001'))]

In [ ]:
#Count savings accounts by their transaction status to see how many are successful, pending, failed, etc.

%%sql
SELECT
    transaction_status,
    COUNT(*) AS count
FROM
    savings_savingsaccount
GROUP BY
    transaction_status;

 * mysql+pymysql://root:***@localhost/adashi_staging
27 rows affected.


[('success', 106727),
 ('failed', 54144),
 ('Charge attempt cannot be fulfilled until Sat Apr 25 2020 16:38:51 GMT+0000 (UTC) (284 minutes)', 1),
 ('Pending', 74),
 ('reward', 31),
 ('Expired card', 1),
 ('redemption', 1059),
 ('Charge attempt cannot be fulfilled until Wed Apr 29 2020 09:42:16 GMT+0000 (UTC) (163 minutes)', 1),
 ('earnings', 925),
 ('monnify_success', 321),
 ('Transaction amount limit exceeded', 1),
 ('Charge attempt cannot be fulfilled until Sun Apr 26 2020 21:32:02 GMT+0000 (UTC) (1437 minutes)', 1),
 ('circle', 30),
 ('Charge attempt cannot be fulfilled until Sun Apr 26 2020 21:32:02 GMT+0000 (UTC) (747 minutes)', 1),
 ('Charge attempt cannot be fulfilled until Wed Apr 29 2020 09:42:16 GMT+0000 (UTC) (1427 minutes)', 1),
 ('Charge attempt cannot be fulfilled until Sun Apr 26 2020 21:32:02 GMT+0000 (UTC) (317 minutes)', 1),
 ('Charge attempt cannot be fulfilled until Sun Apr 26 2020 21:32:02 GMT+0000 (UTC) (270 minutes)', 1),
 ('Charge attempt cannot be fulfilled until Wed Apr 29 2020 09:42:16 GMT+0000 (UTC) (238 minutes)', 1),
 ('successful', 2),
 ('support credit', 2),
 ('supportcredit', 74),
 ('abandoned', 26),
 ('reversal', 266),
 ('usd_index_redemption', 11),
 ('New Card Initialization Redemption.', 1),
 ('failed_new_card_initialization_to_naira_stash_redemption', 32),
 ('Format is Authorization: Bearer [secret key]', 1)]

In [ ]:
# How status varies over time

%%sql
SELECT
    DATE_FORMAT(transaction_date, '%Y-%m') AS month,
    transaction_status,
    COUNT(*) AS count
FROM
    savings_savingsaccount
GROUP BY
    month, transaction_status
ORDER BY
    month DESC, count DESC;

 * mysql+pymysql://root:***@localhost/adashi_staging
383 rows affected.


[('2025-04', 'success', 244),
 ('2025-04', 'reversal', 19),
 ('2025-04', 'redemption', 12),
 ('2025-04', 'failed', 8),
 ('2025-03', 'success', 70),
 ('2025-03', 'failed', 4),
 ('2025-03', 'reversal', 1),
 ('2025-03', 'redemption', 1),
 ('2025-02', 'success', 391),
 ('2025-02', 'failed', 30),
 ('2025-02', 'redemption', 1),
 ('2025-02', 'reversal', 1),
 ('2025-01', 'success', 47),
 ('2025-01', 'supportcredit', 4),
 ('2025-01', 'failed', 3),
 ('2025-01', 'redemption', 1),
 ('2024-12', 'success', 242),
 ('2024-12', 'failed', 11),
 ('2024-12', 'redemption', 6),
 ('2024-12', 'supportcredit', 4),
 ('2024-12', 'reversal', 1),
 ('2024-12', 'earnings', 1),
 ('2024-11', 'success', 129),
 ('2024-11', 'failed', 30),
 ('2024-11', 'reversal', 6),
 ('2024-11', 'redemption', 4),
 ('2024-10', 'redemption', 649),
 ('2024-10', 'success', 64),
 ('2024-10', 'reversal', 2),
 ('2024-10', 'failed', 1),
 ('2024-09', 'failed', 1852),
 ('2024-09', 'success', 1088),
 ('2024-09', 'reversal', 3),
 ('2024-09', 'redemption', 1),
 ('2024-08', 'failed', 8992),
 ('2024-08', 'success', 5562),
 ('2024-08', 'reversal', 17),
 ('2024-08', 'supportcredit', 11),
 ('2024-08', 'redemption', 4),
 ('2024-07', 'failed', 7517),
 ('2024-07', 'success', 5547),
 ('2024-07', 'redemption', 5),
 ('2024-07', 'reversal', 3),
 ('2024-07', 'supportcredit', 1),
 ('2024-06', 'failed', 6236),
 ('2024-06', 'success', 2082),
 ('2024-06', 'reversal', 19),
 ('2024-06', 'redemption', 1),
 ('2024-05', 'failed', 10383),
 ('2024-05', 'success', 3405),
 ('2024-05', 'reversal', 4),
 ('2024-05', 'redemption', 3),
 ('2024-04', 'failed', 4422),
 ('2024-04', 'success', 1210),
 ('2024-04', 'supportcredit', 26),
 ('2024-04', 'redemption', 1),
 ('2024-03', 'failed', 1733),
 ('2024-03', 'success', 879),
 ('2024-03', 'redemption', 6),
 ('2024-03', 'reversal', 4),
 ('2024-02', 'failed', 7579),
 ('2024-02', 'success', 2134),
 ('2024-02', 'reversal', 10),
 ('2024-02', 'supportcredit', 2),
 ('2024-02', 'redemption', 2),
 ('2024-01', 'supportcredit', 9),
 ('2024-01', 'redemption', 8),
 ('2024-01', 'success', 6),
 ('2023-12', 'success', 8443),
 ('2023-12', 'failed', 463),
 ('2023-12', 'redemption', 7),
 ('2023-11', 'success', 9379),
 ('2023-11', 'failed', 448),
 ('2023-11', 'redemption', 16),
 ('2023-11', 'reversal', 1),
 ('2023-10', 'success', 7616),
 ('2023-10', 'failed', 647),
 ('2023-10', 'redemption', 32),
 ('2023-10', 'reversal', 1),
 ('2023-09', 'success', 5162),
 ('2023-09', 'failed', 763),
 ('2023-09', 'redemption', 4),
 ('2023-09', 'reversal', 1),
 ('2023-08', 'success', 4884),
 ('2023-08', 'failed', 285),
 ('2023-08', 'earnings', 173),
 ('2023-08', 'redemption', 44),
 ('2023-08', 'reversal', 4),
 ('2023-07', 'success', 3413),
 ('2023-07', 'failed', 207),
 ('2023-07', 'earnings', 42),
 ('2023-07', 'redemption', 11),
 ('2023-07', 'reversal', 3),
 ('2023-06', 'success', 5602),
 ('2023-06', 'failed', 509),
 ('2023-06', 'reversal', 124),
 ('2023-06', 'redemption', 3),
 ('2023-05', 'success', 1415),
 ('2023-05', 'failed', 124),
 ('2023-05', 'reversal', 8),
 ('2023-05', 'usd_index_redemption', 3),
 ('2023-04', 'success', 2053),
 ('2023-04', 'failed', 53),
 ('2023-04', 'reversal', 6),
 ('2023-04', 'redemption', 1),
 ('2023-04', 'Format is Authorization: Bearer [secret key]', 1),
 ('2023-03', 'success', 1628),
 ('2023-03', 'failed', 79),
 ('2023-03', 'failed_new_card_initialization_to_naira_stash_redemption', 4),
 ('2023-03', 'redemption', 4),
 ('2023-03', 'reversal', 1),
 ('2023-02', 'success', 2319),
 ('2023-02', 'failed', 136),
 ('2023-02', 'failed_new_card_initialization_to_naira_stash_redemption', 9),
 ('2023-02', 'monnify_success', 4),
 ('2023-02', 'redemption', 3),
 ('2023-02', 'reversal', 2),
 ('2023-01', 'success', 2260),
 ('2023-01', 'failed', 109),
 ('2023-01', 'failed_new_card_initialization_to_naira_stash_redemption', 15),
 ('2023-01', 'redemption', 5),
 ('2022-12', 'success', 2154),
 ('2022-12', 'failed', 87),
 ('2022-12', 'monnify_success', 22),
 ('2022-12', 'failed_new_card_initialization_to_na

In [88]:
#Analyzing fees paid
%%sql
SELECT
    SUM(fee_in_cents)/100 AS total_fees_paid,
    AVG(fee_in_cents)/100 AS avg_fee_per_transaction
FROM
    savings_savingsaccount;

 * mysql+pymysql://root:***@localhost/adashi_staging
1 rows affected.


[(17916.18, 0.10942114134948942)]

In [ ]:
#Number of savings accounts created per month

%%sql
SELECT
    DATE_FORMAT(created_on, '%Y-%m') AS month,
    COUNT(*) AS accounts_created
FROM
    savings_savingsaccount
GROUP BY
    month
ORDER BY
    accounts_created DESC
LIMIT 12;

 * mysql+pymysql://root:***@localhost/adashi_staging
12 rows affected.


[('2024-08', 14586),
 ('2024-05', 13795),
 ('2024-07', 13073),
 ('2023-11', 9844),
 ('2024-02', 9728),
 ('2023-12', 8913),
 ('2024-06', 8338),
 ('2023-10', 8296),
 ('2023-06', 6238),
 ('2023-09', 5930),
 ('2024-04', 5662),
 ('2023-08', 5390)]

In [87]:
#Average Deduction vs. Confirmed Amount To see how much is deducted vs. confirmed on average

%%sql
SELECT
    AVG(deduction_amount) AS avg_deduction,
    AVG(confirmed_amount) AS avg_confirmed
FROM savings_savingsaccount


 * mysql+pymysql://root:***@localhost/adashi_staging
1 rows affected.


[(3270805.5543230497, 3324660.357055367)]

In [ ]:

#Creating a mapping to group the statuses
%%sql
SELECT
  CASE
    WHEN transaction_status IN ('success', 'monnify_success', 'successful') THEN 'Success'
    WHEN transaction_status = 'failed' THEN 'Failed'
    WHEN transaction_status LIKE 'Charge attempt%' THEN 'System Message'
    WHEN transaction_status IN ('reward', 'redemption', 'earnings', 'circle') THEN 'Internal Movement'
    WHEN transaction_status IN ('support credit', 'supportcredit') THEN 'Support Credit'
    WHEN transaction_status IN ('pending', 'reversal', 'abandoned') THEN UPPER(transaction_status)
    ELSE 'Other'
  END AS normalized_status,
  COUNT(*) AS status_count
FROM
  savings_savingsaccount
GROUP BY
  normalized_status
ORDER BY
  status_count DESC;

 * mysql+pymysql://root:***@localhost/adashi_staging
9 rows affected.


[('Success', 107050),
 ('Failed', 54144),
 ('Internal Movement', 2045),
 ('REVERSAL', 266),
 ('Support Credit', 76),
 ('PENDING', 74),
 ('Other', 47),
 ('ABANDONED', 26),
 ('System Message', 8)]

In [ ]:
%%sql SELECT * FROM users_customuser LIMIT 5;

 * mysql+pymysql://root:***@localhost/adashi_staging
5 rows affected.


[('pbkdf2_sha256$216000$YfnYATX59yTt$O0Yum7CxiuJokITPrSiuZlNlyoLYF2lmQ/Pwtf7YeTY=', datetime.datetime(2023, 6, 17, 10, 49), 0, '00372ecec6e449c595fb44b7a5f2c9c0', 'esther+isd@cowrywise.com', None, 'Esther', 'Omikunle', '+234 706 293 2370', None, 0, 1, datetime.datetime(2023, 6, 17, 10, 43, 11), 0, 'esther_isd', datetime.datetime(2023, 6, 17, 10, 43, 11), None, None, None, None, None, 0, None, None, None, 'c1c59bb68b4f81b34410ed91024acde2c50f7fd0cca50859531a5842483dd7c4:88bdb195c0604ba9af5454bb99586285', 'e5fd2b44043d4cdebca762e218226168', 0, 'COWRYWISE', 0, 0, 0, 'COWRYWISE', 1, None, None, None, None, 0.0, 0.0, 0, None, 0, None, None, None, datetime.datetime(2023, 6, 17, 10, 43, 32, 423468), 0, None, 0, 0, None, None, None, 'web', None, None),
 ('pbkdf2_sha256$216000$1zy4LDCtaX5r$KwmEuEFd78lBNzwwoDxWiDuGwNRE1ZHRSVZJ+21XdfY=', datetime.datetime(2023, 3, 19, 21, 48), 0, '0043d851435740f59180a27a136899dc', 'chinonso+69@cowrywise.com', None, 'Garuba', 'Sigma', '+234 814 951 0435', datetime.date(2023, 3, 25), 0, 1, datetime.datetime(2023, 3, 19, 21, 47, 46), 0, 'chinonso_69', datetime.datetime(2023, 3, 19, 21, 47, 46), 2, None, None, None, None, 0, None, None, None, '3064b4f717ce3852edbab67c114728f5b33c81e52cf1944cb524146383f70b23:412e63508f85489687aef49e3bc91cd5', 'a4c62f30096d42338627c8cbb6dfcc09', 0, 'COWRYWISE', 0, 0, 0, 'COWRYWISE', 1, None, None, None, None, 0.0, 0.0, 0, None, 0, None, None, None, datetime.datetime(2023, 3, 23, 4, 58, 29, 161295), 0, None, 0, 0, None, None, None, 'web', None, None),
 ('pbkdf2_sha256$320000$1mE7Q9WXF1rfH9ezOaxBDe$sCjtt4U2jc564bhPPMckuMbVbqSYlwA7YZjJwUyzmcA=', datetime.datetime(2025, 3, 9, 16, 11, 4), 0, '005320d615be4bcd8a18a1631406cc86', 'yami+109@cowrywise.com', None, 'Yami109', 'Jeremiah', '+234 802 376 7919', None, 0, 1, datetime.datetime(2025, 3, 9, 16, 7, 30), 0, 'yami_109', datetime.datetime(2025, 3, 9, 16, 7, 30), None, None, None, None, None, 0, None, None, None, '39f24eab71fc35013c057b1915fde8e89d75efe724b2f9578b94709f74a0bccb:437f708cd1f14851bca1a789fc9472ba', 'f27f3a449ea544b4bc0b49738209f37a', 0, 'COWRYWISE', 0, 0, 0, 'COWRYWISE', 0, None, None, None, None, 0.0, 0.0, 0, None, 0, None, None, None, datetime.datetime(2025, 3, 9, 16, 7, 41, 702317), 0, None, 0, 0, None, None, None, 'web', None, None),
 ('pbkdf2_sha256$216000$EyY1gc5gW1Is$ZxauL//F5ceIl4gy4DMiJqrz1Z8L8Ub21pP4yo4aaho=', datetime.datetime(2023, 10, 14, 16, 8, 6), 0, '00603278f1494e19aa0403a6ac97806f', 'taslim+12938123@cowrywise.com', None, 'Tas', 'Lim', '+234 706 197 9046', None, 0, 1, datetime.datetime(2023, 10, 14, 16, 6, 36), 0, 'taslim12938123', datetime.datetime(2023, 10, 14, 16, 6, 36), None, None, None, None, None, 0, None, None, None, '0b152e57f46ac27317690d4ed91391ca45ac1f4344338f457e18b5304aa64083:35a09cad514e45238e3290021d89d3f9', None, 0, 'google-play', 0, 0, 0, 'COWRYWISE', 0, None, None, None, None, 0.0, 0.0, 0, None, 0, None, 'organic', None, datetime.datetime(2023, 10, 14, 16, 6, 56, 631285), 0, None, 0, 0, None, None, None, 'android', None, None),
 ('pbkdf2_sha256$216000$muYiJHIHO8Fy$u3aTpT0jAolIIGGHML4CB7sYKcsTGE87Vp2AJOC4ank=', datetime.datetime(2024, 12, 28, 9, 25, 35), 0, '0081f354fdc54a83abf66579629eba34', 'lanre+test7@cowrywise.com', None, 'Timothy', 'Olanrewaju', '+234 816 299 4545', datetime.date(2012, 5, 25), 0, 1, datetime.datetime(2021, 5, 25, 11, 19, 57), 0, 'lanretest7', datetime.datetime(2021, 5, 25, 11, 19, 57), 2, '', None, None, None, 2, '37.4219983', '-122.084', '', 'f2b118e793f703df04b5aca4d5b0ec253eb1c980f0a816f430a351f3cb406df5:d2275998b61b4448890c4b2ac3cfba41', 'a98663187e4d4b61b7de96a5e8547871', 0, 'COWRYWISE', 0, 0, 0, 'COWRYWISE', 0, None, None, None, None, 0.0, 50000000.0, 0, None, 0, None, None, None, datetime.datetime(2021, 5, 25, 11, 20, 19, 815117), 0, None, 0, 0, None, None, None, None, None, 3)]

In [90]:
#Users with no savings activity
%%sql
SELECT u.id, u.email
FROM users_customuser u
LEFT JOIN savings_savingsaccount s ON u.id = s.owner_id
WHERE s.id IS NULL;

 * mysql+pymysql://root:***@localhost/adashi_staging
994 rows affected.


[('aae6bc85b49d4e4aa6d625738e7bd989', 'a@admin.com'),
 ('8ff0c8a0d3ab4c3a9ee7fb58447591ea', 'aclpool+james@gmail.com'),
 ('f35a6dec41eb43aab5bd924ff48702b6', 'aclpool+koko11@cowrywise.com'),
 ('d920e45a23c345e9bb92c17aba85b42d', 'aclpool+koko12@cowrywise.com'),
 ('aea9a6163b2042baaacb78070bcf8cf9', 'aclpool+koko13@cowrywise.com'),
 ('ddd26ea905d94b819bb7325d8d183f2b', 'aclpool+koko14@cowrywise.com'),
 ('6363c58649b04224ba4eae4628b4a35b', 'aclpool+koko15@cowrywise.com'),
 ('55d188417f424459ba3222bbb90882ba', 'aclpool+master@gmail.com'),
 ('a5bf3dbcf8674fe994e42d8ab9b81e8f', 'aclpool+segun@gmail.com'),
 ('ae53a225ca9a4dbbb7834ac6cedec154', 'aclpool+testnum@cowrywise.com'),
 ('6504479ec51641b9bfe608ed7c3adad3', 'aclpool+threes@gmail.comxxxxxxxxx'),
 ('2a8524fa8a7f4a46b296b5eb3dc5b1d8', 'adimchi+20@cowrywise.com'),
 ('0c8bc6dcf1d94f93975a399c1784ad24', 'adimchi+2@cowrywise.com'),
 ('d0060a3ae5c441559b84da37dc8c3fbb', 'adimchi+3@cowrywise.com'),
 ('576530ee0ebf4585a506e8fcc89ecd25', 'adimchi+6@cowrywise.com'),
 ('bfadd92623ad44e98eb00978f03c4205', 'afoo+test10@cowrywise.com'),
 ('3f9c10f4f34e4c12b79e131c8f8219fe', 'afoo+test11@cowrywise.com'),
 ('75061e844ecd40f5a8c1096c2df8edf0', 'afoo+test12@cowrywise.com'),
 ('1d1e1dc8685945e2b2dddc49803e2dcb', 'afoo+test13@cowrywise.com'),
 ('a94ce5891733488a9428c9a1d1254172', 'afoo+test14@cowrywise.com'),
 ('acd06f68a88947c88786a8d52e08202e', 'afoo+test6@cowrywise.com'),
 ('655cf5b1155b46aba2e0a1c3299b01c8', 'afoo+test7@cowrywise.com'),
 ('806fba4574484bc9b5d488aa48f4320d', 'afoo+test9@cowrywise.com'),
 ('11af8c455bf747a49e7115425f84c5ee', 'afoo@cowrywise.com'),
 ('57bb571ce4ff441ab251f3b14ad37c57', 'ahmed+1000+daniella@cowrywise.com'),
 ('f25780b3cfa34dae9252083db744940a', 'ahmed+1000+david@cowrywise.com'),
 ('96c55ec2cb7e406d9bed0f24a7e4bca8', 'ahmed+1000+eliana@cowrywise.com'),
 ('70147a2b69bc44ba936442951271caa0', 'aisha+11@cowrywise.com'),
 ('aba81eb463484cbfa8cf5d754406d47c', 'aisha+1@cowrywise.com'),
 ('95f4a650ebed4fd0a4d630eebf6e9f1e', 'aisha+2@cowrywise.com'),
 ('3b29d0ddf3b64a51b5688510eb7ca45f', 'aisha+3@cowrywise.com'),
 ('c03820b795bb4383a98fb298601e8d09', 'anagboguchinonso97@gmail.com'),
 ('86e5ff30c74b464bae482a1b7efdc043', 'anagboguchristopher@gmail.com'),
 ('569a2e67ed4c42aab26312a2422b846c', 'anakin@cowrywise.com'),
 ('156dcb9f229d46b3bec1802c14094b0b', 'aquila4christ@gmail.com'),
 ('a3c05cca7dda444ba3fb19072116620d', 'ayo16345@gmail.com'),
 ('819278f49937410e986577e2218486d2', 'b@admin.com'),
 ('591ef0168af044b6838e07a491b807bc', 'barakat@cowrywise.com'),
 ('d561bd5987ad4b3a9bd99829b73ea102', 'blessing+test@cowrywise.com'),
 ('46d74027d6a044909f900ddf623cf58d', 'blessing@cowrywise.com'),
 ('8265e84236f54847a2daccadbd8517e1', 'boluwatifesamuel48@gmail.com'),
 ('fba7f27812e4425698d06f08ce99b3d9', 'c@admin.com'),
 ('bb7c9d604fa64dcbb7070bb082c10956', 'chigozie+aaron@cowrywise.com'),
 ('4b671eb5677e4c808b8ed1f756e87532', 'chima+99@cowrywise.com'),
 ('d3edd528c8c540aeb7fa0ff5a7ed8d91', 'chima+ads@cowrywise.com'),
 ('57774e330e39468d9e8f1a6169ffbeb2', 'chima+super@cowrywise.com'),
 ('1c5dbc333ab3442e97daf332c20abc30', 'chinonso+1009@cowrywise.com'),
 ('dabd9b95106942d7bd9fe0a236eb1aa4', 'chinonso+100@cowrywise.com'),
 ('bbd13fb280a544eab39ec89559928afd', 'chinonso+101@cowrywise.com'),
 ('8699bb479c8641199957cf0ca2a5aae9', 'chinonso+102@cowrywise.com'),
 ('7bb2bb3761af4ef5b8d54ed0c21620d5', 'chinonso+103@cowrywise.com'),
 ('d88641bd4b274ed3ba11f330ef50758a', 'chinonso+104@cowrywise.com'),
 ('267a2e5b5a8b463ca7ff5ed55370262d', 'chinonso+105@cowrywise.com'),
 ('98c822c89d5e4bbc8f4d0cc5fc73679f', 'chinonso+108@cowrywise.com'),
 ('b09d5ed40c164a0ca4f7d6ae4c442d96', 'chinonso+110@cowrywise.com'),
 ('0df01f83e7bd4b09ac2cdba7869d93fc', 'chinonso+111@cowrywise.com'),
 ('196105f43a29424d924d73bb3d741eb2', 'chinonso+112@cowrywise.com'),
 ('335724a35af24535acd98913ca58d0f7', 'chinonso+113@cowrywise.com'),
 ('a66022753bb046a9b0a6b69266a59608', 'chinonso+114@cowrywise.com'),
 ('d1bd13ec7b9846e

In [97]:
#Number of Active vs Inactive Users
%%sql
SELECT
    is_active,
    COUNT(*) AS count
FROM users_customuser
GROUP BY is_active;

 * mysql+pymysql://root:***@localhost/adashi_staging
2 rows affected.


[(1, 1689), (0, 178)]

In [95]:
%%sql SELECT * FROM withdrawals_withdrawal LIMIT 5;

 * mysql+pymysql://root:***@localhost/adashi_staging
5 rows affected.


[(1, 10000.0, 10000.0, 'W52366', datetime.datetime(2016, 8, 31, 21, 23, 8), 4.38356164379911, 1, '0257625a02344b239b41e1cbe60ef080', '671e6e5caf0f463b80123c839264de8c', 1, 2, 4, 0.0, '', None, None, None, 'NGN', 0.0, None, datetime.datetime(2023, 6, 13, 16, 21, 57, 25419), datetime.datetime(2023, 6, 13, 16, 21, 57, 195922), None),
 (2, 1500000.0, 0.0, 'W56184', datetime.datetime(2016, 9, 3, 16, 40, 18), 40018.0821917808, 1, '0257625a02344b239b41e1cbe60ef080', '671e6e5caf0f463b80123c839264de8c', 1, 1, 4, 0.0, '', None, None, None, 'NGN', 0.0, None, datetime.datetime(2023, 6, 13, 16, 21, 57, 25419), datetime.datetime(2023, 6, 13, 16, 21, 57, 195922), None),
 (3, 1300000.0, 0.0, 'W56739', datetime.datetime(2016, 9, 3, 16, 42, 25), 40018.0821917808, 1, '0257625a02344b239b41e1cbe60ef080', '671e6e5caf0f463b80123c839264de8c', 1, 1, 4, 0.0, '', None, None, None, 'NGN', 0.0, None, datetime.datetime(2023, 6, 13, 16, 21, 57, 25419), datetime.datetime(2023, 6, 13, 16, 21, 57, 195922), None),
 (4, 10000.0, 10000.0, 'W61100', datetime.datetime(2016, 12, 11, 10, 9, 20), 32967590.6641956, 10, '0257625a02344b239b41e1cbe60ef080', '7f960e5b3743434882ee8ddcedb9aecb', 1, 2, 4, 0.0, '', None, None, None, 'NGN', 0.0, None, datetime.datetime(2023, 6, 13, 16, 21, 57, 25419), datetime.datetime(2023, 6, 13, 16, 21, 57, 195922), None),
 (5, 12000.0, 12000.0, 'W72359', datetime.datetime(2016, 12, 11, 10, 36, 51), 32955590.6641956, 10, '0257625a02344b239b41e1cbe60ef080', '7f960e5b3743434882ee8ddcedb9aecb', 1, 2, 4, 0.0, '', None, None, None, 'NGN', 0.0, None, datetime.datetime(2023, 6, 13, 16, 21, 57, 25419), datetime.datetime(2023, 6, 13, 16, 21, 57, 195922), None)]

In [84]:
#Total Transaction Volume
%%sql
SELECT
    SUM(amount) AS total_transaction_amount
FROM withdrawals_withdrawal;


 * mysql+pymysql://root:***@localhost/adashi_staging
1 rows affected.


[(22462323657.68,)]

In [85]:
#Total Amount Withdrawn
%%sql
SELECT
    SUM(amount_withdrawn) AS total_withdrawn
FROM withdrawals_withdrawal;


 * mysql+pymysql://root:***@localhost/adashi_staging
1 rows affected.


[(22363655657.68,)]

In [91]:
#Total number of users
%%sql
SELECT COUNT(*) AS total_users FROM users_customuser;

 * mysql+pymysql://root:***@localhost/adashi_staging
1 rows affected.


[(1867,)]

In [93]:
#Signup Sources
%%sql
SELECT
    account_source,
    COUNT(*) AS user_count
FROM users_customuser
GROUP BY account_source
ORDER BY user_count DESC;

 * mysql+pymysql://root:***@localhost/adashi_staging
14 rows affected.


[('COWRYWISE', 1489),
 ('google-play', 306),
 ('GTCO', 37),
 ('email', 7),
 ('web', 6),
 ('hni-signup', 5),
 ('gift_redemption', 5),
 ('listing', 3),
 ('in_app', 3),
 ('mtn', 2),
 ('cw', 1),
 ('test', 1),
 ('hr', 1),
 ('customer.io', 1)]

In [94]:
#Most Common Signup Device
%%sql
SELECT
    signup_device,
    COUNT(*) AS device_count
FROM users_customuser
GROUP BY signup_device
ORDER BY device_count DESC;

 * mysql+pymysql://root:***@localhost/adashi_staging
4 rows affected.


[('web', 703), (None, 554), ('android', 326), ('ios', 284)]